### PR TITLE
Reword monitoring federation recommendation

### PR DESCRIPTION
**Pull request by pecigonzalo**
We would like to be more explicit about our recommendation **not** to federate our metrics.

### CODE REVIEW

File: doc/admin/observability/metrics.md [hunk 1]
```diff
@@ -104,13 +104,11 @@ More behavior can be controlled with
 
 ### FAQ
 
-#### Can I consume Sourcegraph's Prometheus metrics in my own monitoring system (Datadog, New Relic, etc.)?
+#### Can I consume Sourcegraph's metrics in my own monitoring system (Datadog, New Relic, etc.)?
 
-It is technically possible to consume all of Sourcegraph's Prometheus metrics in any external monitoring system that supports Prometheus scraping (both Datadog and New Relic support this). However, we would advise against it because Sourcegraph is a very complex system and defining all of the alerting thresholds and rules that are needed to ensure Sourcegraph is healthy is very tedious and changes with each release of Sourcegraph.
+While It's technically possible to consume all of Sourcegraph's metrics in any external system, our recommended approach is to configure Sourcegraph to send [alerts to your own PagerDuty, Slack, email, etc.](alerting.md) and utilize the built in systems to view the system health.
```
Comment (bobheadxi) [inline doc/admin/observability/metrics.md:109]:
```suggestion
While it is technically possible to consume all of Sourcegraph's metrics in any external system, our recommended approach is to configure Sourcegraph to send [alerts to your own PagerDuty, Slack, email, etc.](alerting.md) and utilize the built in systems to view system health.
```

File: doc/admin/observability/metrics.md [hunk 2]
```diff
@@ -104,13 +104,11 @@ More behavior can be controlled with
...
+The builtin dashboards and alerts provided by Sourcegraph are updated on each release to assure that the information you are monitoring is up-to-date. Defining all of the alerting thresholds and rules that are needed to ensure Sourcegraph is healthy manually would be very tedious as metrics and thresholds can change with each release of Sourcegraph.
```
Comment (bobheadxi) [inline doc/admin/observability/metrics.md:110]:
```suggestion
The [builtin dashboards and alerts provided by Sourcegraph](dashboards.md) are updated on each release to assure that the information you are monitoring is up-to-date. Defining all of the alerting thresholds and rules that are needed to ensure Sourcegraph is healthy manually would be very tedious as metrics and thresholds can change with each release of Sourcegraph.
```

File: doc/admin/observability/metrics.md [hunk 3]
```diff
@@ -104,13 +104,11 @@ More behavior can be controlled with
...
 
-Most commonly [Sourcegraph's monitoring is configured to send alerts to your own PagerDuty, Slack, email, etc.](alerting.md). Less common approaches include:
-
-- Using [the HTTP API to query which alerts are firing in Sourcegraph](alerting_custom_consumption.md) in order to pipe that information into your own monitoring platform.
-- Using [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) to pull in Sourcegraph's metrics into your own Prometheus instance, which includes our [high-level alerting metrics](metrics_guide.md) which we define via baked-in Prometheus rules.
+External monitoring system that supports Prometheus scraping (both Datadog and New Relic support this) or [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) can be used to consume Sourcegraph's metrics, which includes our [high-level alerting metrics](metrics_guide.md).
```
Comment (bobheadxi) [inline doc/admin/observability/metrics.md:113]:
```suggestion
External monitoring systems that support Prometheus scraping (for example, Datadog and New Relic) or [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) can be used to consume Sourcegraph's metrics, which includes our [high-level alerting metrics](metrics_guide.md).
```

Comment (bobheadxi) [reply doc/admin/observability/metrics.md:113]: (maybe repeat the recommendation here as well?)

**Review by bobheadxi — APPROVED** [review]

**Comment by slimsag** [comment]
General feedback: the rewording here makes me think "I cannot consume Sourcegraph's metrics in my own monitoring system at all" which would encourage me to push Sourcegraph to support this. However, if you read closely we are saying you can consume our _high level metrics_ which I think is good enough for most people (I can name a few customers using this). Can we rephrase to more clearly suggest this in the first paragraph ("do not try to consume our individual metrics, but we have high-level ones you can consume" or something along those lines?)

File: doc/admin/observability/metrics.md (hunk 2 above)
Comment (pecigonzalo) [reply doc/admin/observability/metrics.md]: I prefer to keep the link a bit shorter I expanded to `[send alerts ...`

**Comment by pecigonzalo** [comment]
> the rewording here makes me think "I cannot consume Sourcegraph's metrics in my own monitoring system at all"

@slimsag That is the intention, as far as I know we dont support or document "federating" the metrics (via Prometheus or otherwise).

- I could not find any docs on "how to setup a system to pull those metrics", only how to query them via API https://docs.sourcegraph.com/admin/observability/metrics_guide. 
- I believe even those high level metrics are not stable over time and they are added/removed ad-hoc without a changelog note.

File: doc/admin/observability/metrics.md [hunk 4]
```diff
@@ -104,13 +104,9 @@ More behavior can be controlled with
...
-
-Most commonly [Sourcegraph's monitoring is configured to send alerts to your own PagerDuty, Slack, email, etc.](alerting.md). Less common approaches include:
-
-- Using [the HTTP API to query which alerts are firing in Sourcegraph](alerting_custom_consumption.md) in order to pipe that information into your own monitoring platform.
-- Using [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) to pull in Sourcegraph's metrics into your own Prometheus instance, which includes our [high-level alerting metrics](metrics_guide.md) which we define via baked-in Prometheus rules.
+External monitoring systems that support Prometheus scraping (for example, Datadog and New Relic) or [Prometheus federation](https://prometheus.io/docs/prometheus/latest/federation/) can be used to consume Sourcegraph's metrics, which includes our [high-level alerting metrics](metrics_guide.md).
```
Comment (slimsag) [inline doc/admin/observability/metrics.md:111]: @pecigonzalo we link to a document about them here

**Comment by slimsag** [comment]
> as far as I know we dont support or document "federating" the metrics (via Prometheus or otherwise).

we have historically suggested this to customers whenever they ask about prometheus federation etc. and pointed them to https://docs.sourcegraph.com/admin/observability/metrics_guide#high-level-alerting-metrics - I agree we don't have any docs in place for _how_ to configure Prometheus federation, though.

https://app.hubspot.com/contacts/2762526/company/407948923/ is one customer consuming these (through the API).

> I believe even those high level metrics are not stable over time and they are added/removed ad-hoc without a changelog note.

I don't think this is the case, we have historically kept these stable over time intentionally for https://app.hubspot.com/contacts/2762526/company/407948923/ - with one caveat: that _services_ in this metric can be added/removed ad-hoc without a changelog note. (i.e. if we get rid of gitserver, its metrics are gone - but that would be obvious because the k8s YAML changelog would note we removed the service)

File: doc/admin/observability/metrics.md (hunk 4 above)
Comment (pecigonzalo) [reply doc/admin/observability/metrics.md:111]: @slimsag a document about which part specifically? The references on how to use those I found in that document are "To set up notifications for these alerts, see: alerting.", or https://docs.sourcegraph.com/admin/observability/alerting_custom_consumption which only references checking them via `curl`, not federating the metrics.

The alert names are also not described, it only describes the pattern of `alert_count`.

Comment (slimsag) [reply doc/admin/observability/metrics.md:111]: sorry 😅 I was writing my response in the main thread and just pointing this place out

**Comment by pecigonzalo** [comment]
> we have historically suggested this to customers whenever they ask about prometheus federation etc. and pointed them to https://docs.sourcegraph.com/admin/observability/metrics_guide#high-level-alerting-metrics - I agree we don't have any docs in place for how to configure Prometheus federation, though.
> 
> https://app.hubspot.com/contacts/2762526/company/407948923/ is one customer consuming these (through the API).

I understand the old docs highlighting that what you could do with the API was not consume the metrics but rather "use the http API to query which alerts are firing" and Im trying to be more explicit about it. 

Maybe I can move up the "Alternatively, you can leverage the provided " part to be after the main paragraph to provide an alternative to consuming the metrics closer to the "unsupported" bit. Or I can do something like "Sourcegraph provides a single metric over HTTP query `alerts_count` (link to doc)" and then add the "While its technically possible [...]". 🤔 Thoughts?

> I don't think this is the case, we have historically kept these stable over time intentionally for https://app.hubspot.com/contacts/2762526/company/407948923/ - with one caveat: that services in this metric can be added/removed ad-hoc without a changelog note. (i.e. if we get rid of gitserver, its metrics are gone - but that would be obvious because the k8s YAML changelog would note we removed the service)

As I understand only the format `alert_count{level, service_name, name, description` is stable for them, but we dont provide a stable set of names or parameters so a customer could monitor lets say `alert_count(name~=".*goroutines.*"}`.

--- 

Im looking it at this from this particular question "Can I consume Sourcegraph's metrics in my own monitoring system (Datadog, New Relic, etc.)?" which Im almost certain the answer is "While its possible this is not a supported use case" or "No" but that is a bit blunt.

**Comment by slimsag** [comment]
> I understand the old docs highlighting that what you could do with the API was not consume the metrics but rather "use the http API to query which alerts are firing" and Im trying to be more explicit about it.

You are right this is what the old docs are saying. I am merely trying to add that historically anytime someone has asked "can I consume Sourcegraph's metrics in my system?" our answer has been "No, but you can consume our high-level metrics and rely on the structure of those (but not the exact alerts that are firing - i.e. you can use whatever to scrape our metrics and then _only_ alert on the number of open alerts on Sourcegraph's side"

> As I understand only the format `alert_count{level, service_name, name, description` is stable for them, but we dont provide a stable set of names or parameters so a customer could monitor lets say `alert_count(name~=".*goroutines.*"}`.

Correct.

> Im looking it at this from this particular question "Can I consume Sourcegraph's metrics in my own monitoring system (Datadog, New Relic, etc.)?" which Im almost certain the answer is "While its possible this is not a supported use case" or "No" but that is a bit blunt.

I would say the answer is: "not the individual metrics themselves, but you can set up alerts on 'how many alerts are firing in Sourcegraph's system?' so your system notifies you when Sourcegraph is unhealthy and you should look at its dashboards"

> Or I can do something like "Sourcegraph provides a single metric over HTTP query alerts_count (link to doc)" and then add the "While its technically possible [...]".

This sounds okay to me, I just wanted to raise that there is an answer we have used historically besides just "no", which is that you **can** consume `alert_count` (as a count only, not relying on the specific data/labels inside it) and this should help to reduce the amount of customers asking for a definitive set of stable metrics and thresholds ("our monitoring tells us if Sourcegraph is unhealthy at a high-level and we should visit their dashboards to figure out why" is better and more acceptable than "our monitoring has no info about Sourcegraph").

Anyway, sorry for derailing - I'll chime out here and I trust you here :)

**Review by slimsag — APPROVED** [review]
(approving to signal I am in favor of the idea here in general)

**Review by christinelovett — APPROVED** [review]

**Review by daxmc99 — APPROVED** [review]
LGTM

**Comment by pecigonzalo** [comment]
@slimsag No worries, Im trying to find a middle point between in which we are also clear about direct federation not being supported at the moment (given we have no docs or any other "standard" way of doing this). I changed the order and rephrased a bit to try and get it to that point.
- HTTP API supported
- Scraping / Federation possible but not recommended or supported


@daxmc99 @christinelovett could you give this a 2nd look? If the new format its not as clear I can just revert the last commit.

**Comment by davejrt** [comment]
LGTM

**Pull request merged**